dialer_cache.go

```diff
@@ -0,0 +1,170 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudsqlconn
+
+import (
+	"sync"
+
+	"cloud.google.com/go/cloudsqlconn/debug"
+	"cloud.google.com/go/cloudsqlconn/instance"
+)
+
+// dialerCache manages a thread-safe map of ConnName to monitoredCache. This
+// provides thread safe operations to atomically get or add an entry, scan all
+// entries and update or remove some of them, and to find entries using the
+// domain name or ConnName.
+type dialerCache struct {
+	mu     sync.RWMutex
+	cache  map[instance.ConnName]*monitoredCache
+	logger debug.ContextLogger
+}
+
+// newDialerCache creates and initializes an instance of the dialer cache
+func newDialerCache(logger debug.ContextLogger) *dialerCache {
+	return &dialerCache{
+		mu:     sync.RWMutex{},
+		cache:  make(map[instance.ConnName]*monitoredCache),
+		logger: logger,
+	}
+}
+
+// replaceAll thread-safe iterate through all cache entries, replace or removing
+// the entries. f() provides the replacement values
+//   - no change: return cn, c
+//   - replace: return an instance.ConnName and monitoredCache. This will
+//     replace the old entry.
+//   - remove: should return the empty value instance.ConnName{} if the entry
+//     should be removed.
+//
+// This method is not re-entrant.
+func (d *dialerCache) replaceAll(f func(cn instance.ConnName, c *monitoredCache) (instance.ConnName, *monitoredCache)) {
+	emptyInstance := instance.ConnName{}
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	newCache := make(map[instance.ConnName]*monitoredCache)
+	for cn, c := range d.cache {
+		newCn, newC := f(cn, c)
+		// ignore entries that have empty instance names
+		if newCn == emptyInstance {
+			continue
+		}
+		newCache[newCn] = newC
+	}
+	d.cache = newCache
+}
+
+// findByDomainName returns the entry that matches the domain name.
+// dn - the domain name
+// returns:
+//
+//	instance.ConnName the name of the matching instance
+//	monitoredCache the cached item
+//	bool true when there is a result.
+//
+// This method is not re-entrant.
+func (d *dialerCache) findByDomainName(dn string) (instance.ConnName, *monitoredCache, bool) {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+	for cn, c := range d.cache {
+		if cn.DomainName() == dn {
+			return cn, c, true
+		}
+	}
+	return instance.ConnName{}, nil, false
+}
+
+// get returns the instance matching the cn.
+//
+// This method is not re-entrant.
+func (d *dialerCache) get(cn instance.ConnName) (*monitoredCache, bool) {
```

Comment:
https://go.dev/doc/effective_go#Getters

Reply:
This is an indexed accessor like `cache[key]`, not a getter. Other well-known go cache libraries have chosen to name their "get from cache" method `Get(key)`, so we have done the same here.

- [istio Cache](https://pkg.go.dev/istio.io/pkg/cache#Cache)
- [bigcache](https://github.com/allegro/bigcache)
- [go-cache](https://github.com/patrickmn/go-cache)